Convent stm32 iwdg driver to new API

qianfan-Zhao opened the pull request.

qianfan-Zhao committed:
drivers: iwdg_stm32: Convent STM32 watchdong driver to new API

Fixes: #10917

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.com>

qianfan-Zhao committed:
samples: watchdog: Remove stm32 iwdg sample

STM32 iwdg driver are convented to new API now, the old sample
are doesn't need again.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao commented:
@apodogrocki Could you please also review this PR?

erwango commented on drivers/watchdog/iwdg_stm32.c:
```diff
@@ -61,29 +62,35 @@ static void iwdg_stm32_convert_timeout(u32_t timeout,
+		LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_DBGMCU);
+		LL_DBGMCU_APB1_GRP1_FreezePeriph(LL_DBGMCU_APB1_GRP1_IWDG_STOP);
+	}
+
```
Add check for WDT_OPT_PAUSE_IN_SLEEP and return ENOTSUP in that case

galak replied:
Done

erwango commented on drivers/watchdog/iwdg_stm32.c:
```diff
@@ -101,60 +108,39 @@ static int iwdg_stm32_set_config(struct device *dev,
-
-	tickstart = k_uptime_get_32();
-
-	while (LL_IWDG_IsReady(iwdg) == 0) {
```
Why do you remove this part?

apodogrocki replied:
Exactly, this part is present in the HAL stm32f3xx_hal_iwdg.h file in method **HAL_IWDG_Init**. Purpose is clear: Wait for register to be updated. Should stay in my opinion.

qianfan-Zhao replied:
the old version enable IWDG(LL_IWDG_Enable) first and then do iwdg_stm32_set_config. So it's ok to wait IDDG ready.

But the new API doesn't enable IWDG when install timeout, it is enabled when setup. 

I don't think it's ok waiting IWDG ready even if it is doesn't enabled.

galak commented:
@qianfan-Zhao any updates based on @erwango comments?

qianfan-Zhao commented:
@galak I am doing another work this weekdays and doesn't have enough time to complete this. So please give me some times.😁

galak commented:
> @galak I am doing another work this weekdays and doesn't have enough time to complete this. So please give me some times.😁

No worries, thanks for the update.

apodogrocki commented on drivers/watchdog/iwdg_stm32.c:
```diff
@@ -101,60 +108,39 @@ static int iwdg_stm32_set_config(struct device *dev,
 	/* Neither STM32F1X nor STM32F4 series supports window option. */
 	LL_IWDG_SetWindow(iwdg, 0x0FFF);
```
It is a little bit confusing. Currently there is no option added in the Kconfig to activate Window Mode. So setting it to the default value 0x0000 0FFF in fact does nothing. It neither modifies Window register IWDG_WINR nor reloads the Reload Register IWDG_RLR.
Maybe it would be worth to add appropriate config option to Kconfig.stm32?

galak replied:
Would it be good to just remove this for now, and someone can add the feature in the future?

apodogrocki commented:
@qianfan-Zhao approval refers to the second commit. Must have pressed something too quickly.

galak commented:
@qianfan-Zhao just checking in to see were we are on this.  Thanks.

erwango commented:
@qianfan-Zhao , Can you take into account https://github.com/zephyrproject-rtos/zephyr/issues/12069 in your driver?

erwango commented:
@vaussard, I can't add you to the review, but it would be valuable you have a check if you have some time.

vaussard commented:
Hi @erwango, I will see what I can do, but do not wait on me.

galak commented:
@qianfan-Zhao any updates on this?

galak commented:
> Hi @erwango, I will see what I can do, but do not wait on me.

@vaussard got any time to look at this again?

erwango committed:
soc: stm32: Extended watchdog support to all series

STM32F2, STM32F7 and STM32L0 were missing wtachdog API support.
Fix this.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: stm32: Add watchdog nodes to STM32 dtsi files

Provide watchdog node definition to stm32 dtsi files to enable
watchdog configuration by device tree.
Add matching st,stm32-watchdog binding.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/watchdog: stm32: move driver to dts based configuration

STM32 watchdog driver is now configured from device tree settings.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32: watchdog enable watchdog support with dt

Now that stm32 watchdog should be configured by device tree,
update dts file of boards declaring watchdog support.
Additionally update doc and yaml files.
Add support on some boards that were used to validate the driver
update:
- disco_l475_iot1
- nucleo_f207zg
- nucleo_f429zi
- nucleo_f746zg
- nucleo_f073rz


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/watchdog: stm32: Disable IWDG_STM32_START_AT_BOOT by default

STM32 specific behavior "IWDG_STM32_START_AT_BOOT" is enabled
by default. As all vendor specific behavior, this is preferred
to be disabled by default and activated at application level,
so application can have better control of the driver.
For instance watchdog test can't run with this setting.
Disable by default.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests: watchdog: Update test to support drivers with no callback

STM32 watchdog driver does not support callback which is not expected
by the test. Update to make test able to run on STM32 drivers.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@galak, as discussed
